# Implementation of Pearson's chi squared test

In [1]:
import pandas as pd
import numpy as np
from scipy import stats as ss

In [2]:
hr_data = pd.read_csv('HRDataset_v9.csv')
hr_data.head()

Employee Name  Employee Number  MarriedID  MaritalStatusID  \
0            Brown, Mia       1103024456          1                1   
1  LaRotonda, William         1106026572          0                2   
2      Steans, Tyrone         1302053333          0                0   
3       Howard, Estelle       1211050782          1                1   
4           Singh, Nan        1307059817          0                0   

   GenderID  EmpStatus_ID  DeptID  Perf_ScoreID  Age  Pay Rate  ...  \
0         0             1       1             3   30     28.50  ...   
1         1             1       1             3   34     23.00  ...   
2         1             1       1             3   31     29.00  ...   
3         0             1       1             9   32     21.50  ...   
4         0             1       1             9   30     16.56  ...   

  Date of Hire  Days Employed Date of Termination       Reason For Term  \
0   10/27/2008           3317                 NaN  N/A - still employed   
1     1/6/2014           1420                 NaN  N/A - still employed   
2    9/29/2014           1154                 NaN  N/A - still employed   
3    2/16/2015             58           4/15/2015  N/A - still employed   
4     5/1/2015            940                 NaN  N/A - still employed   

  Employment Status     Department                  Position  \
0            Active  Admin Offices              Accountant I   
1            Active  Admin Offices              Accountant I   
2            Active  Admin Offices              Accountant I   
3            Active  Admin Offices  Administrative Assistant   
4            Active  Admin Offices  Administrative Assistant   

         Manager Name         Employee Source         Performance Score  
0  Brandon R. LeBlanc      Diversity Job Fair               Fully Meets  
1  Brandon R. LeBlanc      Website Banner Ads               Fully Meets  
2  Brandon R. LeBlanc         Internet Search               Fully Meets  
3  Brandon R. LeBlanc  Pay Per Click - Google  N/A- too early to review  
4  Brandon R. LeBlanc      Website Banner Ads  N/A- too early to review  

[5 rows x 28 columns]

### Contigency table

In [3]:
cross = pd.crosstab(hr_data['Manager Name'], hr_data['Performance Score'])
cross

Performance Score   90-day meets  Exceeds  Exceptional  Fully Meets  \
Manager Name                                                          
Alex Sweetwater                2        1            1            4   
Amy Dunn                       1        2            1           11   
Board of Directors             0        0            0            2   
Brandon R. LeBlanc             1        0            0            4   
Brannon Miller                 1        5            2            8   
Brian Champaigne               0        0            0            8   
David Stanley                  1        1            0           15   
Debra Houlihan                 0        0            0            2   
Elijiah Gray                   3        2            0           13   
Eric Dougall                   0        1            0            3   
Janet King                     0        3            1           13   
Jennifer Zamora                0        1            1            4   
John Smith                     1        0            0           11   
Kelley Spirea                  3        3            0           15   
Ketsia Liebig                  2        2            0           14   
Kissy Sullivan                 3        1            1           14   
Lynn Daneault                  1        2            0            8   
Michael Albert                 1        2            0            9   
Peter Monroe                   3        0            0            6   
Simon Roup                     5        0            2            6   
Webster Butler                 3        2            0           11   

Performance Score   N/A- too early to review  Needs Improvement  PIP  
Manager Name                                                          
Alex Sweetwater                            0                  1    0  
Amy Dunn                                   5                  0    1  
Board of Directors                         0                  0    0  
Brandon R. LeBlanc                         2                  0    0  
Brannon Miller                             0                  1    4  
Brian Champaigne                           0                  0    0  
David Stanley                              4                  0    0  
Debra Houlihan                             0                  1    0  
Elijiah Gray                               3                  1    0  
Eric Dougall                               0                  0    0  
Janet King                                 0                  2    0  
Jennifer Zamora                            0                  1    0  
John Smith                                 0                  1    1  
Kelley Spirea                              1                  0    0  
Ketsia Liebig                              2                  1    0  
Kissy Sullivan                             1                  1    1  
Lynn Daneault                              1                  0    1  
Michael Albert                             6                  3    1  
Peter Monroe                               5                  0    0  
Simon Roup                                 4                  0    0  
Webster Butler                             3                  2    0

In [4]:
cross.columns

Index(['90-day meets', 'Exceeds', 'Exceptional', 'Fully Meets',
       'N/A- too early to review', 'Needs Improvement', 'PIP'],
      dtype='object', name='Performance Score')

In [6]:
cross = cross.drop('N/A- too early to review', 1)
cross

Performance Score   90-day meets  Exceeds  Exceptional  Fully Meets  \
Manager Name                                                          
Alex Sweetwater                2        1            1            4   
Amy Dunn                       1        2            1           11   
Board of Directors             0        0            0            2   
Brandon R. LeBlanc             1        0            0            4   
Brannon Miller                 1        5            2            8   
Brian Champaigne               0        0            0            8   
David Stanley                  1        1            0           15   
Debra Houlihan                 0        0            0            2   
Elijiah Gray                   3        2            0           13   
Eric Dougall                   0        1            0            3   
Janet King                     0        3            1           13   
Jennifer Zamora                0        1            1            4   
John Smith                     1        0            0           11   
Kelley Spirea                  3        3            0           15   
Ketsia Liebig                  2        2            0           14   
Kissy Sullivan                 3        1            1           14   
Lynn Daneault                  1        2            0            8   
Michael Albert                 1        2            0            9   
Peter Monroe                   3        0            0            6   
Simon Roup                     5        0            2            6   
Webster Butler                 3        2            0           11   

Performance Score   Needs Improvement  PIP  
Manager Name                                
Alex Sweetwater                     1    0  
Amy Dunn                            0    1  
Board of Directors                  0    0  
Brandon R. LeBlanc                  0    0  
Brannon Miller                      1    4  
Brian Champaigne                    0    0  
David Stanley                       0    0  
Debra Houlihan                      1    0  
Elijiah Gray                        1    0  
Eric Dougall                        0    0  
Janet King                          2    0  
Jennifer Zamora                     1    0  
John Smith                          1    1  
Kelley Spirea                       0    0  
Ketsia Liebig                       1    0  
Kissy Sullivan                      1    1  
Lynn Daneault                       0    1  
Michael Albert                      3    1  
Peter Monroe                        0    0  
Simon Roup                          0    0  
Webster Butler                      2    0

In [10]:
cross.shape

(21, 6)

In [7]:
ss.chi2_contingency(cross)

(104.19464939881554,
 0.3670945067181692,
 100,
 array([[ 1.02197802,  0.92307692,  0.2967033 ,  5.96703297,  0.49450549,
          0.2967033 ],
        [ 1.81684982,  1.64102564,  0.52747253, 10.60805861,  0.87912088,
          0.52747253],
        [ 0.22710623,  0.20512821,  0.06593407,  1.32600733,  0.10989011,
          0.06593407],
        [ 0.56776557,  0.51282051,  0.16483516,  3.31501832,  0.27472527,
          0.16483516],
        [ 2.38461538,  2.15384615,  0.69230769, 13.92307692,  1.15384615,
          0.69230769],
        [ 0.90842491,  0.82051282,  0.26373626,  5.3040293 ,  0.43956044,
          0.26373626],
        [ 1.93040293,  1.74358974,  0.56043956, 11.27106227,  0.93406593,
          0.56043956],
        [ 0.34065934,  0.30769231,  0.0989011 ,  1.98901099,  0.16483516,
          0.0989011 ],
        [ 2.15750916,  1.94871795,  0.62637363, 12.5970696 ,  1.04395604,
          0.62637363],
        [ 0.45421245,  0.41025641,  0.13186813,  2.65201465,  0.21978022,
     

 $$ chi\ squared\ test\ statistics = 104.19464939881554$$<br>
 $$p\ value\ =\ 0.3670945067181692$$<br>
 $$degree\ of\ freedom,(dof)\ =\ 100$$
> The array represents the Expected Outcome<br>
> Since the `p-value` is greater than $0.05$, we fail to reject the null hypothesis and conclude that there is no relationship between the Performance score and Manager Name (who an employee works for).

In [8]:
def cramers_corrected_stat(confusion_matrix):
    """ calculate Cramers V statistic for categorial-categorial association.
        uses correction from Bergsma and Wicher, 
        Journal of the Korean Statistical Society 42 (2013): 323-328
    """
    chi2 = ss.chi2_contingency(confusion_matrix)[0]
    n = confusion_matrix.sum().sum()
    phi2 = chi2/n
    r,k = confusion_matrix.shape
    phi2corr = max(0, phi2 - ((k-1)*(r-1))/(n-1))    
    rcorr = r - ((r-1)**2)/(n-1)
    kcorr = k - ((k-1)**2)/(n-1)
    return np.sqrt(phi2corr / min( (kcorr-1), (rcorr-1)))

In [11]:
cramers_corrected_stat(cross)

0.053443126803626846

> The cramer's V value is closer to zero(0), meaning there is little to no relationship between the two variables.